vllm/entrypoints/openai/api_server.py

```diff
@@ -520,8 +520,11 @@ async def run_server(args, **uvicorn_kwargs) -> None:
     logger.info("vLLM API server version %s", VLLM_VERSION)
     logger.info("args: %s", args)
 
-    temp_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-    temp_socket.bind(("", args.port))
```

Comment:
out of curiosity, I still don't understand why this does not work when frontend multiprocessing is disabled.

Reply:
@youkaichao 

Creating the engine causes the executor to be instantiated and initialized in a new process using `multiprocessing`.

Doing this after the socket is created and bound, causes a reference to the socket being passed onto the child, causing resources not to be freed properly.

Here's a minimal example that reproduces the issue:


```python
import asyncio
import uvloop
import socket
import uvicorn
import multiprocessing
import time

from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI()

mp_method = "fork"
# mp_method = "spawn"
mp = multiprocessing.get_context(mp_method)


class Request(BaseModel):
    text: str


@app.post("/")
async def echo(request: Request):
    return {"request": request.text}


def worker(name: str, *args):
    """dummy worker"""
    time.sleep(5)


def spawn_process(*args):
    process = mp.Process(
        target=worker,
        name="worker",
        args=("process_worker", *args),
        # kwargs={}
        daemon=True,
    )
    process.start()
    return process


HOST, PORT = "0.0.0.0", 8000


async def serve_http():
    config = uvicorn.Config(
        app=app,
        host=HOST,
        port=PORT,
        timeout_keep_alive=5,
    )
    server = uvicorn.Server(config)

    loop = asyncio.get_running_loop()
    server_task = loop.create_task(server.serve())

    await server_task


async def run_server():
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

    s.bind((HOST, PORT))

    process = spawn_process() # in vllm, this block is `build_async_engine_client`
    print(f"spawned {process=}")

    s.close()

    await serve_http()

    process.join()


if __name__ == "__main__":
    uvloop.run(run_server())

```